Add support for encrypted CA and certificate private keys

PeterFeicht opened the pull request:
I've been working on adding support for encrypted private keys, I've added support to the CA and certificate managers, as well as the user manager. I'm working on the OpenVPN wizard now, but wanted to get some feedback on what I have so far. I'm not a PHP programmer at all, so I just tried to stick to the existing code; any feedback in that regard is more than welcome.

I've not added support to the CRL manager yet, because I think that would require a substantial change to its interface. I'd be happy to discuss that, but I wasn't sure if a PR was the right place. Since I don't think I'm going to use CRLs at all, I could also just leave that part out for someone else to implement, if the need is there.

- [x] Redmine Issue: https://redmine.pfsense.org/issues/1257
- [x] Ready for review

PeterFeicht commented on src/etc/inc/certs.inc:
```diff
@@ -238,15 +241,19 @@ function ca_create(& $ca, $keylen, $lifetime, $dn, $digest_alg = "sha256") {
 
 	$signing_ca_res_crt = openssl_x509_read(base64_decode($signing_ca['crt']));
-	$signing_ca_res_key = openssl_pkey_get_private(array(0 => base64_decode($signing_ca['prv']) , 1 => ""));
+	$signing_ca_res_key = openssl_pkey_get_private(array(0 => base64_decode($signing_ca['prv']) , 1 => $capass));
```
Why are the arguments passed as an array to `openssl_pkey_get_private`, does that have any significance?

PeterFeicht replied:
Will do.

PeterFeicht commented on src/usr/local/www/system_camanager.php:
```diff
@@ -99,6 +99,7 @@
 	$pconfig['serial'] = $a_ca[$id]['serial'];
 	if (!empty($a_ca[$id]['prv'])) {
 		$pconfig['key'] = base64_decode($a_ca[$id]['prv']);
+		$pconfig['encryptkey'] = (strpos($pconfig['key'], 'ENCRYPTED') !== false);
```
I don't know if that's the right way to check for encrypted keys. It was this way before, but there's the very slim chance that the base64 encoded string might contain `ENCRYPTED`. Maybe this should be a function in `certs.inc`?

jim-p replied:
IIRC you can try the openssl function `openssl_pkey_get_private()` on the key with an empty password and if it fails/returns empty then it's safe to assume that it's encrypted. I haven't tested that so I'm not sure exactly what/how it returns, but that should be reliable.

PeterFeicht replied:
I'll try that, and I'll factor the check out into a separate function in any case.

PeterFeicht commented on src/usr/local/www/system_usermanager.php:
```diff
@@ -705,7 +721,7 @@ function build_cert_table() {
 		null,
 		'hidden',
-		''
+		$act
```
I think this should have been `$act` before, but please have a look if it's right the way it is now.

jim-p replied:
It should be empty as it was.

PeterFeicht replied:
But then there's the problem that the empty POST parameter overrides the non-empty GET parameter, and the form won't show right when you submit but there's an error.

PeterFeicht commented on src/usr/local/www/system_camanager.php:
```diff
@@ -616,6 +688,11 @@
 	$pconfig['caref'],
 	$allCas
 ));
+$group->add(new Form_Input(
```
Should this be in a separate section? I couldn't find a way to make it look nice when both are toggled, so I added the password field to the group.

jim-p replied:
Have to check it out when testing to be sure but it's probably OK like this

PeterFeicht replied:
👍 It looks fine on desktop, it even flows nicely on small screens.

PeterFeicht commented:
Added a few comments to the diff.

jim-p commented on src/etc/inc/certs.inc:
```diff
@@ -622,6 +647,40 @@ function cert_get_publickey($str_crt, $decode = true, $type = "crt") {
+	file_put_contents($keyfn, $str_key);
+	file_put_contents($passfn, $password);
+
+	exec("/usr/bin/openssl pkey -in {$keyfn} -des3 -passout file:{$passfn} 2>/dev/null", $out);
```
Why 3DES (`-des3`) here, and not something more secure, like `-aes128` or `-aes256`?

PeterFeicht replied:
I just always used `-des3`, I don't have a preference either way.

jim-p replied:
I'd prefer to see this as at least `-aes128`, otherwise this looks OK to merge and test.

jim-p commented on src/etc/inc/certs.inc:
```diff
@@ -622,6 +647,40 @@ function cert_get_publickey($str_crt, $decode = true, $type = "crt") {
 			break;
 	}
 	unlink($certfn);
+	return ($out !== false ? implode("\n", $out) : false);
```
The comparison here should be inside parenthesis, so `return (($out !== false) ? implode("\n", $out) : false);` for example

PeterFeicht replied:
ACK

jim-p commented on src/etc/inc/certs.inc:
```diff
@@ -601,15 +614,27 @@ function cert_get_issuer($str_crt, $decode = true) {
+
+				exec("/usr/bin/openssl pkey -in {$certfn} -passin file:{$passfn} -pubout 2>/dev/null", $out, $status);
+
+				unlink($passfn);
```
It's safer to use `unlink_if_exists()` here instead of `unlink()`. Will be the same for other uses throughout this code.

PeterFeicht replied:
Should I go ahead and change all uses of `unlink`?

jim-p commented on src/etc/inc/certs.inc:
```diff
@@ -622,6 +647,40 @@ function cert_get_publickey($str_crt, $decode = true, $type = "crt") {
+	}
+
+	$keyfn = tempnam('/tmp', 'CDK');
+	$passfn = tempnam('/tmp', 'CDK');
```
Same as above, use unique prefixes.

jim-p commented on src/usr/local/www/system_certmanager.php:
```diff
@@ -195,6 +195,20 @@
 
 if ($_POST['save']) {
+	// If passwords are given, check that they match and verify their length
+	function validatePasswords(& $input_errors, $pw1, $pw2) {
```
Why not put this function somewhere common (like `certs.inc`, or `guiconfig.inc`) and use it here and on `system_camanager.php`? It would need a more specific name, though, such as `validate_cert_passwords()`.

PeterFeicht replied:
That's not a bad idea considering that it'll be needed for the OpenVPN wizard too.

PeterFeicht commented on src/etc/inc/certs.inc:
```diff
@@ -601,15 +614,27 @@ function cert_get_issuer($str_crt, $decode = true) {
 		case 'prv':
-			exec("/usr/bin/openssl pkey -in {$certfn} -pubout", $out);
+			if ($password) {
+				$passfn = tempnam('/tmp', 'CGPK');
```
Though the documentation says unique, apparently there have been problems. I'll change it.

PeterFeicht commented:
I've thought about that too, I agree there needs to be a check at least. Since certificates are already filtered in such lists based on whether they even have a private key, I'll just add another check there whether the key is encrypted. I'll add a corresponding check as well.

It would also be nice to have some kind of indication whether a certificate/CA has an encrypted private key in the respective lists. I don't quite know where to put that, a separate column would make the lists a little too crowded I think. For the certificate list I'm thinking a padlock symbol after the purpose string maybe? For CAs the same after the name?

Regarding encrypted export, since encrypted keys are exported as they are anyway, I'd prefer to leave that for a separate PR.

PeterFeicht commented:
I finally got around to finishing up this PR.

 * Added a new function `get_cert_list` for getting a list of certificates (including or excluding encrypted ones), which is used in most places that display a list of certificates. There are checks against encrypted certificates at those places too.
 * The OpenVPN wizard now supports creation of new certificates using an encrypted CA; encrypted certificates are filtered from all wizard certificate lists (currently only used in the OpenVPN wizard).
 * I also added an indication for encrypted private keys to the CA and certificate lists.

There's only one open question: https://github.com/pfsense/pfsense/pull/4062#discussion_r274039904

jim-p commented:
I didn't see that reply. I guess we can try populating that, as long as it doesn't negatively affect existing behavior.

PeterFeicht commented:
Just had a quick look and I think it needs to be set in other places too. I'll get back to you when I've had some time to look it over in detail.

rbgarga commented:
@PeterFeicht do you still have plans to work on this PR?

rbgarga commented:
If yes, please rebase your fork. The change is conflicting and cannot be merged

PeterFeicht commented:
Yeah, sorry for the long silence, I'm still on it. I'll have a look at it in the next couple of days I think.

When I rebase my branch, would it make sense to work the review changes into my commits to clean up the branch, or how do guys usually handle this?

rbgarga commented:
If you can squash your commits and split them in commits that make sense is better since we don't use the squash function from github's GUI.

PeterFeicht commented:
So I just rebased onto the master branch and squashed review changes, the diff didn't change at all.
I'll use fixup commits for future review changes (since I know how to use those now, really handy I'll tell you) and squash the branch when it's ready to be merged.

I still have to look at the `$act` thing in detail.

rbgarga commented:
> So I just rebased onto the master branch and squashed review changes, the diff didn't change at all.
> I'll use fixup commits for future review changes (since I know how to use those now, really handy I'll tell you) and squash the branch when it's ready to be merged.
> 
> I still have to look at the `$act` thing in detail.

great. I'll wait to hear from you about `$act` change and then test it

PeterFeicht commented:
I've had a look at the `$act` thing now and I'm pretty sure it's fine the way it is now (setting the POST `act` variable to `$act` in [system_usermanager.php:753](https://github.com/pfsense/pfsense/pull/4062/files#diff-adf66c0efb366e75ef2329bafa9d239bR753)).

During my testing I discovered two other things that didn't work quite right and I fixed them too. Please let me know if I should file an issue and move those commits to a separate branch/PR instead.
I also removed a redundant if statement, this could go in a separate PR as well.

Let me know if/when you're ready to merge the branch, then I will squash that fixup commit one last time.

rbgarga commented on src/usr/local/www/system_usermanager.php:
```diff
@@ -742,307 +742,306 @@ function build_cert_table() {
 	exit;
 }
 
+// ($act == "new" || $act == "edit" || $input_errors) == true
```
Please remove this commented out line

PeterFeicht replied:
Done.

rbgarga commented:
@PeterFeicht do you still plan to work on it?

PeterFeicht commented:
@rbgarga yes I do. I'm sorry for the long delay, I haven't been well lately.
I should be able to work on it again in the next couple of weeks. This currently targets 2.5, right? Do you have any estimate on when that will be? I'm asking because I tend to get things done better with a deadline looming ;-)

rbgarga commented:
@PeterFeicht do you still have plans to work on this change?

PeterFeicht commented:
Yes I do, sorry for the long inactivity.

rbgarga commented:
Closing due to inactivity since Sep 2020

Dantali0n commented:
This is very disappointing to see, I was looking forward to this new feature :'(
